### DISCUSSION TITLE
Nilan EC9 - User function switch and modbus question

### DISCUSSION THREAD

**Discussion by ImsRims** [veista/nilan#69] [Q&A]
Hi,

First of all, thanks for the great integration! I'm new to this topic and I couldn't solve this one by myself.

Am I able to switch user function (in finnish etäkytkinkäyttö/takkakytkin) ON or OFF via this integration alone or do I need separate switch (which to control) to do that? I would like to switch user function on/off remotely via HA when I start fire in fireplace.

Other thing, I don't have in my configuration panel anything related to floor warming or heating curve. Should these be there? My modbus version seems to be 11, if that helps. I checked other discussions and I assume that 22 is the latest. How can I update my modbus version? There is picture attached.

Thanks in advance!
![Sieppaa](https://github.com/veista/nilan/assets/139852480/9d63e2b7-4a16-4692-8331-a36fb12cc0d3)

**Comment by veista** [comment]
User function: Well to my knowledge the user functions can only be triggered via physical switches. Though you can achieve the same functionality via this integration.

Floor warming and heating curve: I'm not sure. They could be there by mistake. Most compressor models do have heating curve setting.

Update: You can read it on the wiki page.

**Reply by ImsRims** [reply]
Thanks for answer. I made button in HA and wrote a code to configuration.yaml which should change the state of user function. Here's the code:

switch:
  - platform: template
    switches:
      nilan_kytkin:
        friendly_name: "Takkakytkin"
        value_template: "{{ is_state('binary_sensor.nilan_program_user_func_act', '1') }}"
        turn_on:
          service: modbus.write_register
          data_template:
            unit: 30
            address: 600
            value: 1
        turn_off:
          service: modbus.write_register
          data_template:
            unit: 30
            address: 600
            value: 0

I'm not very good at writing a code, but HA doesn't complain about this. Can you see if there is some values or something that needs to be changed or if something is missing?

When I press the button in dashboard I get this message:

![kuva](https://github.com/veista/nilan/assets/139852480/f2643576-4023-4141-b21c-1daa0ce72d46)

Related to heating curve items, what do you mean by they could be there by mistake? I meant that there are nothing related to heating curve settings in configuration panel which came originally with this integration.

**Reply by veista** [reply]
Like I said before, I'm not aware that the user function can be accessed via software. You need a physical switch to turn those on. I am not able to help you with that.

I have mapped all the available functions to each model from the user manuals available from Nilan. There might be mistakes on your model. There also might be differences between manufacturing years etc.
